src/pykeen/datasets/ea/combination.py

```diff
@@ -44,69 +182,151 @@ def __call__(
             a single triples factory comprising the joint graph, as well as a tensor of pairs of matching IDs.
             The tensor of matching pairs has shape `(2, num_alignments)`, where `num_alignments` can also be 0.
         """
+        # concatenate triples
+        mapped_triples, offsets = cat_triples(left, right)
+        # filter alignment and translate to IDs
+        alignment = filter_map_alignment(alignment=alignment, left=left, right=right, entity_offsets=offsets[:, 0])
+        # process
+        mapped_triples, alignment, translation_kwargs = self._process(mapped_triples, alignment, offsets)
+        if isinstance(left, TriplesFactory) and isinstance(right, TriplesFactory):
+            # merge mappings
+            entity_to_id, relation_to_id = _merge_both_mappings(
+                # TODO: offsets can also be dense mapping
+                # translation: Union[torch.LongTensor, Tuple[torch.LongTensor, torch.LongTensor]]
+                left=left,
+                right=right,
+                **translation_kwargs,
+            )
+            triples_factory = TriplesFactory(
+                mapped_triples=mapped_triples,
+                entity_to_id=entity_to_id,
+                relation_to_id=relation_to_id,
+                **kwargs,
+            )
+        else:
+            triples_factory = CoreTriplesFactory(mapped_triples=mapped_triples, **kwargs)
+
+        return triples_factory, alignment
+
+    @abstractmethod
+    def _process(
+        self,
+        mapped_triples: MappedTriples,
+        alignment: torch.LongTensor,
+        offsets: torch.LongTensor,
+    ) -> Tuple[MappedTriples, torch.LongTensor, Mapping[str, Any]]:
         raise NotImplementedError
 
 
+class DisjointGraphPairCombinator(GraphPairCombinator):
+    """This combinator keeps both graphs as disconnected components."""
+
+    # docstr-coverage: inherited
+    def _process(
+        self,
+        mapped_triples: MappedTriples,
+        alignment: torch.LongTensor,
+        offsets: torch.LongTensor,
+    ) -> Tuple[MappedTriples, torch.LongTensor, Mapping[str, Any]]:  # noqa: D102
+        return mapped_triples, alignment, dict(offsets=offsets)
+
+
+class SwapGraphPairCombinator(GraphPairCombinator):
+    """Add extra triples by swapping aligned entities."""
+
+    # docstr-coverage: inherited
+    def _process(
+        self,
+        mapped_triples: MappedTriples,
+        alignment: torch.LongTensor,
+        offsets: torch.LongTensor,
+    ) -> Tuple[MappedTriples, torch.LongTensor, Mapping[str, Any]]:  # noqa: D102
+        # add swap triples
+        # e1 ~ e2 => (e1, r, t) ~> (e2, r, t), or (h, r, e1) ~> (h, r, e2)
+        dense_map = torch.full(size=(offsets[-1, 0],), fill_value=-1)
```

Comment:
Maybe we are thinking differently, but I think dense_map should be
`dense_map = torch.full(size=(mapped_triples.max() + 1,), fill_value=-1)`
where then the aligned entities are swapped with the respective ids, else holding `-1`s
And then `_swap_index` should be:
```
def _swap_index(mapped_triples: MappedTriples, dense_map: torch.LongTensor, index: TargetColumn) -> MappedTriples:
    trans = dense_map[mapped_triples[:, index]]
    mask = trans >= 0
    mapped_triples = mapped_triples[mask].clone()
    mapped_triples[:, index] = trans[mask] # here is the change
    return mapped_triples
```
At least with this I get the intended additional triples

Reply:
```python
dense_map = torch.full(size=(mapped_triples.max() + 1,), fill_value=-1)
```
should be equivalent to
```python
dense_map = torch.full(size=(offsets[-1, 0],), fill_value=-1) 
```
since `offsets[:, [0, 2]]` contains the entity offsets in increasing order => `offsets[-1, 0]` is the total number of entities in the unified graph.

https://github.com/pykeen/pykeen/blob/7bad48e8252f21d9d8e690f5f0b599dbb9d87d14/src/pykeen/datasets/ea/combination.py#L247-L249
then fills the swapped entities in, leaving `-1` for all non-aligned entities.

So essentially, the difference is that you would change
https://github.com/pykeen/pykeen/blob/7bad48e8252f21d9d8e690f5f0b599dbb9d87d14/src/pykeen/datasets/ea/combination.py#L151-L156
to
```python
def _swap_index(mapped_triples: MappedTriples, dense_map: torch.LongTensor, index: TargetColumn) -> MappedTriples:
    trans = dense_map[mapped_triples[:, index]]
    mask = trans >= 0
    mapped_triples = mapped_triples[mask].clone()
    mapped_triples[:, index] = trans[mask] # here is the change
    return mapped_triples
```
right? I am wondering why the previous code did work, if there is a shape mismatch :sweat_smile:

Reply:
done in 7b8d9650ab606af63bbc788335bee3c834814524

Reply:
Previously it didn't work, because the offset was just the size of the first dataset so I got an index error when setting `dense_map[right_id] = left_id`, but I see you fixed it in https://github.com/pykeen/pykeen/commit/7b8d9650ab606af63bbc788335bee3c834814524 and now everything works as expected :+1: